README.md
# hamt_go

A Go implementation of the HAMT data structure.

A Hash Array Mapped Trie ([HAMT][bagwell2001])
provides fast and memory-efficient access to large amounts of data held
in memory.  Values are stored by **key**.  All HAMT keys are mapped into
fixed length unsigned integers. In this implementation these are 64 bits
long, Go `uint64`s.  The **value** may be of any type but typically is a
pointer to a data structure of interest.

The HAMT trie is essentially a prefix trie where the nodes in the tree
are either tables or leaf entries.

In a simpler implementation at each level there is either a leaf node
or a table with up to `2^w` slots, where w is a small
integer.  The next `w` bits of the hash
derived from the key are construed as an index into that table.  The
table grows dynamically to its maximum number of slots as entries are added.
Whenever there is a
collision at a given level, a pointer to a new table replaces the current
leaf entry,
and both new and old entries move to the child table, possibly
recursing in the process.

The current implementation has been tested with values of
3, 4, 5, and 6 for `w`.  Preliminary performance tests indicate
that `w=6` is optimal.  That is, a table with 64 slots gives the best
performance.

An enhancement introduces a fixed size table at the root.  This is
characterized by another small integer t: the root has 2^t entries
and the table at the root is indexed by the first t bits of the
key's hashcode.  Performance tests show that for optimal performance
the root table should approach the total number of entries in size.

A further enhancement would allow dynamic resizing of the root table.
This has not yet been implemented.

Whereas a normal hash table would be quite large and might periodically
require expensive resizing, the HAMT data structure is roughly
as fast as a hash table, but starts small and consumes more memory only
as needed.

## Limitations

* This code is not thread-safe.  That is, using code must provide any
necessary locking.

* the HAMT algorithm depends upon bit-counting.  On modern Intel and AMD
processors this
can be done using a specific machine-language instruction, POPCNT.  The current
implementation of hamt_go emulates this in software using the
[SWAR][wiki-swar] algorithm,  The emulation code is on the order of ten times
slower than the machine instruction.
*In practice, as measured by the Golang pprof profiler, POPCNT emulation does not significantly slow down computations, because the emulation code simply does not use that many CPU cycles.*
*This is in any case not actually a limitation, but rather an observation: the HAMT alorithm runs faster with hardware support.*

## Project Status

The code works and is reasonably well-tested.
`Insert`and  `Find` operations, while not yet thoroughly optimized,
take approximately  0.5 to 0.6 microseconds each on a lightly-loaded server
(about 1.1us each to insert a million values and verify that the
value can be found using the key, where the root table has `2^18` slots).  As the root table approaches the
number of entries in size, this falls to about 1.0 us, or 500ns/op.

These figures were obtained from *single-threaded* tests.  With some degree
of concurrency, the code runs at about 370 ns/op.

## References

[Bagwell, "Ideal Hash Trees"][bagwell2001]  (2001 PDF)

[Wikipeida, "Hash array mapped trie"][wiki-hamt]

[Wikipedia, "SWAR"][wiki-swar]


[bagwell2001]: http://infoscience.epfl.ch/record/64398/files/idealhashtrees.pdf

[wiki-hamt]: http://en.wikipedia.org/wiki/Hash_array_mapped_trie

[wiki-swar]: http://en.wikipedia.org/wiki/SWAR

## On-line Documentation

More information on the **hamt_go** project can be found
[here](https://jddixon.github.io/hamt_go)


CHANGES
hamt_go/CHANGES

v1.1.9
    2017-11-03
        * correct directory structure, config files                 SLOC 2379
v1.1.8
    2017-04-22
        * add projConfig.toml                                       SLOC 2377
v1.1.7
    2016-09-14
        * update license; add profiles                              SLOC 2370
v1.1.6
    2016-06-23
        * add ghpDoc/{css,img}/                                     SLOC 2357
v1.1.5
    2016-05-24
        * add bkp2U, other bash scripts; update .dvcz               SLOC 1807
v1.1.4
    2016-02-24
        * add countem, update LICENSE                               SLOC 1693
v1.1.3
    2015-12-30
        * fix dead link                                             SLOC 1688
v1.1.2
    2015-11-19
        * minor edits to *.md                                       SLOC 1688
v1.1.1
    2015-06-27
        * update .gitignore                                         SLOC 1688
v1.1.0
    2014-12-12
        * moved BitCount32/64() to xlUtil_go                        SLOC 1692
v1.0.1
    2014-11-06
        * replace magic number 6 with MAX_W                         SLOC 1763
v1.0.0
    2014-10-07
        * switched from pointer receiver to value receiver          SLOC 1762
v0.9.0
    2014-05-20
        * maxDepth() replaced by Table.root plus Root.maxDepth      SLOC 1532
    2014-05-16
        * 5-10% faster with NewTableWithLeaf                        SLOC 1489
v0.8.1
    2014-05-15
        * performance enhancements, then add findLeaf to test loop  SLOC 1446
v0.8.0
    2014-05-12
        * replaced Entry with HTNodeI everywhere; dropped Entry     SLOC 1414
        * added profileHAMT, dropped Table.indices                  SLOC 1463
        * calculated BytesKey.Hashcode() directly                   SLOC 1450
        * Table.MaxSlots(), Table.MaxDepth() functions              SLOC 1453
        * root.insertLeaf(), Table.insertLeaf() optimized           SLOC 1441
v0.7.0
    2014-05-06
        * replace *Entry with NodeI in other tables
        * unique test keys                                          SLOC 1468
    2014-05-10
        * use rnglib_go instead of xlattice_go/rnglib
    2014-05-12
        * corrected rnglib_go import
v0.6.0
    2014-05-06
        * replace *Entry with NodeI in root table                   SLOC 1449
v0.5.0
    2014-05-05
        * unsuccessful find returns nil value, nil error            SLOC 1464
v0.4.0
    2014-04-22
        * introduced root with 2^t entries and no use of POPCOUNT   SLOC 1183
    2014-04-23
        * perf tests debugged                                       SLOC 1248
    2014-04-25
        * added HAMT entry splitting test                           SLOC 1351
        * dupe key handling plus tests; plus cleanup                SLOC 1419
    2014-04-30
        * leaf count tests improved, errors corrected               SLOC 1451
v0.3.0
    2014-04-21
        * debugged HAMT that actually uses SWAR code                SLOC 2793
            - much legacy code still present
    2014-04-22
        * dropped HAMT32, HAMT64, and related code                  SLOC 1046
        * correct handling of t at non-root levels                  SLOC 1049
v0.2.0
    2014-04-17
        * tested, working HAMT64 code                               SLOC 1717
    2014-04-20
        * HAMT32 and HAMT64 replaced by HAMT parameterized by w     SLOC 2827
            and t; legacy code remains, the t parameter is present
            but ignored
v0.1.1
    2014-04-07
        * first level ndx inserts pass test                         SLOC  549
    2014-04-08
        * test verifies first level values correct                  SLOC  571
    2014-04-12
        * tested, working Table32.{insert,find,delete}Entry()       SLOC  730
    2014-04-16
        * recursive Table32 tests succeed                           SLOC  967
        * tested, working HAMT32.{Insert,Find,Delete}               SLOC 1007

v0.1.0
    2014-04-04
        * code extracted from ctries_go testing basic insert algo   SLOC  104
        * added SWAR-based BitCount32/64() as util.go               SLOC  148
        * added Bytes*Key, EntryI, HAMT, Key*I, Leaf* and tests     SLOC  386









table.go
package hamt_go

// hamt_go/table.go

import (
	"bytes"
	"errors"
	"fmt"
	xu "github.com/jddixon/xlUtil_go"
)

var _ = fmt.Print

// This is a non-root table; depth is guaranteed never to be zero.  We
// use a uint64 as a bitmap, with a bit being set representing the fact
// that a slot is in use, so there may not be more than 64 slots, so
// w may not exceed 6 (2^6==64).
type Table struct {
	w      uint // non-root tables have 2^w slots
	t      uint // root table has 2^t slots
	mask   uint64
	bitmap uint64
	slots  []HTNodeI // each nil or a pointer to either a leaf or a table
	root   *Root     // pointer to the fixed-size root table
}

// Debugging / sanity check
func CheckTableParam(depth uint, root *Root) (w, t uint, err error) {
	if root == nil {
		err = NilRoot
	} else {
		w = root.w
		t = root.t
		if w > MAX_W {
			err = MaxTableSizeExceeded
		} else if t+(depth-1)*w > 64 {
			err = MaxTableDepthExceeded
		}
	}
	return
}

func NewTable(depth uint, root *Root) (table *Table, err error) {
	w, t, err := CheckTableParam(depth, root)
	if err == nil {
		table = new(Table)
		table.w = w
		table.t = t
		table.root = root
		flag := uint64(1 << w)
		table.mask = flag - 1
	}
	return
}

// Create a new table and insert a first Leaf into it.
//
func NewTableWithLeaf(depth uint, root *Root, firstLeaf *Leaf) (
	table *Table, err error) {

	w, t, err := CheckTableParam(depth, root)
	if err == nil {
		tbl := new(Table)
		tbl.w = w
		tbl.t = t
		tbl.root = root
		wFlag := uint64(1 << w)
		tbl.mask = wFlag - 1
		shiftCount := t + (depth-1)*w
		hc := firstLeaf.Key.Hashcode() >> shiftCount
		ndx := hc & tbl.mask
		flag := uint64(1 << ndx)
		tbl.slots = []HTNodeI{firstLeaf}
		tbl.bitmap = flag
		if err == nil {
			table = tbl
		}
	}
	return
}

func (table *Table) GetRoot() *Root {
	return table.root
}

// Return the maximum number of slots in the table, 2^w.
func (table *Table) MaxSlots() uint {
	return 1 << table.w
}

// Return a count of leaf nodes in this table.
func (table *Table) getLeafCount() (count uint) {
	for i := 0; i < len(table.slots); i++ {
		node := table.slots[i]
		if node != nil {
			if node.IsLeaf() {
				count++
			} else {
				tDeeper := node.(*Table)
				count += tDeeper.getLeafCount()
			}
		}
	}
	return
}

//
func (table *Table) getTableCount() (count uint) {
	count = 1
	for i := 0; i < len(table.slots); i++ {
		node := table.slots[i]
		if node != nil && !node.IsLeaf() {
			tDeeper := node.(*Table)
			count += tDeeper.getTableCount()
		}
	}
	return
}

//func (table *Table) GetDepth() uint {
//	return uint(table.depth)
//}

// XXX Should modify to remove empty table recursively where this
// is the last entry in the table
func (table *Table) removeFromSlices(offset uint) (err error) {
	curSize := uint(len(table.slots))
	if curSize == 0 {
		err = DeleteFromEmptyTable
	} else if offset >= curSize {
		err = errors.New(fmt.Sprintf(
			"InternalError: delete offset %d but table size %d\n",
			offset, curSize))
	} else if curSize == 1 {
		// XXX LEAVES EMPTY TABLE, WHICH WILL HARM PERFORMANCE
		table.slots = table.slots[0:0]
	} else if offset == 0 {
		table.slots = table.slots[1:]
	} else if offset == curSize-1 {
		table.slots = table.slots[0:offset]
	} else {
		// Get rid of expensive appends.
		// shorterSlots := table.slots[0:offset]
		// shorterSlots = append(shorterSlots, table.slots[offset+1:]...)
		shorterSlots := make([]HTNodeI, curSize-1)
		copy(shorterSlots[0:offset], table.slots[0:offset])
		copy(shorterSlots[offset:], table.slots[offset+1:])
		table.slots = shorterSlots
	}
	return
}

// Enter with hc the hashcode for the key shifted appropriately for the
// current depth, so that the first w bits of the shifted hashcode can
// be used as the index of the leaf in the table.
//
// The caller guarantees that depth <= Root.maxTableDepth.
func (table *Table) deleteLeaf(hc uint64, depth uint, key KeyI) (
	err error) {

	if len(table.slots) == 0 {
		err = NotFound
	} else {
		ndx := hc & table.mask
		flag := uint64(1 << ndx)
		mask := flag - 1
		if table.bitmap&flag == 0 {
			err = NotFound
		} else {
			// the node is present; get its position in the slice
			var slotNbr uint
			if mask != 0 {
				slotNbr = xu.BitCount64(table.bitmap & mask)
			}
			node := table.slots[slotNbr]
			if node.IsLeaf() {
				myLeaf := node.(*Leaf)
				myKey := myLeaf.Key.(BytesKey)
				searchKey := key.(BytesKey)
				if bytes.Equal(searchKey.Slice, myKey.Slice) {
					err = table.removeFromSlices(slotNbr)
					table.bitmap &= ^flag
				} else {
					err = NotFound
				}
			} else {
				// node is a table, so recurse
				depth++
				if depth > table.root.maxTableDepth {
					err = NotFound
				} else {
					tDeeper := node.(*Table)
					hc >>= table.w
					err = tDeeper.deleteLeaf(hc, depth, key)
				}
			}
		}
	}
	return
}

// Enter with hc the hashcode for the key shifted appropriately for the
// current depth, the depth as a zero-based integer, and the full key.
// Return nil if no matching entry is found or the value associated with
// the matching entry or any error encountered.
//
// The caller guarantees that depth<=Root.maxTableDepth.
//
func (table *Table) findLeaf(hc uint64, depth uint, key KeyI) (
	value interface{}, err error) { // 1 of 90 samples cum

	ndx := hc & table.mask // 27 of 52; MOVQ 10(DX),CX
	flag := uint64(1 << ndx)

	if table.bitmap&flag != 0 {
		// the node is present; get its position in the slice
		var slotNbr uint
		mask := flag - 1
		if mask != 0 {
			slotNbr = uint(xu.BitCount64(table.bitmap & mask)) // gets expanded inline; 0/52
		}
		node := table.slots[slotNbr] // 20 of 52 - ADDQ BP,BX
		if node.IsLeaf() {
			myLeaf := node.(*Leaf)
			myKey := myLeaf.Key.(BytesKey) // 5 of 92 cum - lib call assest
			searchKey := key.(BytesKey)    // 7 of 92 cum - lib call assert
			if bytes.Equal(searchKey.Slice, myKey.Slice) {
				value = myLeaf.Value
			}
			// otherwise the value returned is nil
		} else {
			// node is a table, so recurse
			depth++
			if depth <= table.root.maxTableDepth {
				tDeeper := node.(*Table)
				hc >>= table.w
				value, err = tDeeper.findLeaf(hc, depth, key)
			}
			// otherwise the value returned is nil
		}
	}
	return
}

// Enter with hc having been shifted so that the first w bits are ndx.
// 2014-05-13: Performance of this function was considerably improved (runtime
// down 25-50%) by replacing slice appends with slice make/copy sequences.
//
// The caller guarantees that depth <= Root.maxTableDepth.
func (table *Table) insertLeaf(hc uint64, depth uint, leaf *Leaf) (err error) {

	var slotNbr uint // whatever is in first line: about 15 of 37
	ndx := hc & table.mask
	flag := uint64(1 << ndx)
	mask := flag - 1
	if mask != 0 {
		slotNbr = xu.BitCount64(table.bitmap & mask)
	}
	sliceSize := uint(len(table.slots))
	if sliceSize == 0 {
		table.slots = []HTNodeI{leaf}
		table.bitmap |= flag
	} else {
		// Is there is already something at this slotNbr ?
		if table.bitmap&flag != 0 {
			entry := table.slots[slotNbr]

			if entry.IsLeaf() {
				// if it's a leaf, we replace the value iff the keys match
				curLeaf := entry.(*Leaf)
				curKey := curLeaf.Key.(BytesKey)
				newKey := leaf.Key.(BytesKey)
				if bytes.Equal(curKey.Slice, newKey.Slice) {
					// the keys match, so we replace the value
					curLeaf.Value = leaf.Value
				} else {
					var (
						tableDeeper *Table
					)
					depth++
					if depth > table.root.maxTableDepth {
						err = MaxTableDepthExceeded
					} else {
						oldLeaf := entry.(*Leaf)
						tableDeeper, err = NewTableWithLeaf(
							depth, table.root, oldLeaf)
						if err == nil {
							hc >>= table.w // this is hashcode for the NEW leaf
							// then put the new leaf in the new table
							err = tableDeeper.insertLeaf(hc, depth, leaf)
							if err == nil {
								// the new table replaces the existing leaf
								table.slots[slotNbr] = tableDeeper
							}
						}
					}
				}
			} else {
				depth++
				if depth > table.root.maxTableDepth {
					err = MaxTableDepthExceeded
				} else {
					// otherwise it's a table, so recurse
					tDeeper := entry.(*Table)
					hc >>= table.w
					err = tDeeper.insertLeaf(hc, depth, leaf)
				}
			}
		} else if slotNbr == 0 {
			leftSlots := make([]HTNodeI, sliceSize+1)
			leftSlots[0] = leaf
			copy(leftSlots[1:], table.slots[:])
			table.slots = leftSlots
			table.bitmap |= flag
		} else if slotNbr == sliceSize {
			table.slots = append(table.slots, leaf)
			table.bitmap |= flag
		} else {
			leftSlots := make([]HTNodeI, sliceSize+1)
			copy(leftSlots[:slotNbr], table.slots[:slotNbr])
			leftSlots[slotNbr] = leaf
			copy(leftSlots[slotNbr+1:], table.slots[slotNbr:])
			table.slots = leftSlots
			table.bitmap |= flag
		}
	}
	return
}

func (table *Table) IsLeaf() bool {
	return false
}
